File: acmicpc/A02805.java
package acmicpc;

import java.io.*;
import java.util.*;

public class A02805 {
    public static long solution(int[] trees, int required) {
        long low = 0L, high = 2_000_000_000L;
        while (low <= high) {
            long cuttingHeight = low + (high - low) / 2, remainders = 0L;
            for (int tree : trees)
                remainders += tree > cuttingHeight ? tree - cuttingHeight : 0L;
            if (remainders < required)
                high = cuttingHeight - 1L;
            else
                low = cuttingHeight + 1L;
        }
        return high;
    }

    public void reader() throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        int trees[] = new int[n];
        for (int i = 0; i < n; i++)
            trees[i] = Integer.parseInt(st.nextToken());
        System.out.println(solution(trees, m));
    }

    public static void main(String[] args) throws NumberFormatException, IOException {
        A02805 test = new A02805();
        test.reader();
    }
}

File: acmicpc/A02805.py
from sys import stdin as s


def binary_search(trees, required):
    low, high = 0, int(2e9)
    while low <= high:
        cutting_h = (low + high) // 2
        remainders = sum(t - cutting_h for t in trees if t > cutting_h)
        if remainders < required:
            high = cutting_h - 1
        else:
            low = cutting_h + 1
    print(high)


def reader():
    n, m = map(int, s.readline().split())
    trees = list(map(int, s.readline().split()))
    binary_search(trees, m)


if __name__ == '__main__':
    reader()
